### ISSUE TITLE
More control over user invitations.

### ISSUE THREAD

**Issue by egypturnash** [mastodon/mastodon#26102]
### Pitch

At present, I can set one thing for a user class regarding invitations: can they generate them, or not?

I would like finer-grained control over this. 

I would like to be able to make choices like 'this class of user can invite [this many] people over the course of [this many] [days|weeks|months|years]", instead of the current situation of "this user class can either generate no invitations, or can generate an infinite-use invitation as many times as they feel like hitting the 'generate invite link' button".

I do not even want *me* to be able to generate infinite-use invitations, to be honest. 

### Motivation

I feel like any admin running an invitation-only instance would benefit from this.

I trust most of my users to exercise good judgement in inviting their friends, and would love to let them do this in a controlled, moderate fashion, but I do *not* want to give them the ability to generate invitations that are good for an infinite number of uses and have no expiration date. Hell, I really don't trust *myself* to generate invitations like that. But that's the *default setting* on the invitation page.

Because of this, at present I have invitations disabled for everyone except myself, and hand out single-use invites when users request it. Which makes the "invited by" field on a user completely useless.

And as a side note I feel like it would be much better if the dropdowns for choosing the number of uses of an invitation and its expiration date defaulted to the *smallest* numbers instead of the largest. Not once in the five and a half years I've been running my server have I felt a need to generate an invitation good for infinite uses. The vast majority have been single-use.

**Comment by renchap** [comment]
Related to https://github.com/mastodon/mastodon/pull/25136

**Comment by egypturnash** [comment]
Yeah basically that except with a better UI than "set some env variables", I am dumb enough to think dealing with running my instance myself is a good idea, but there are a ton of people out there on places like masto.host who don't have that option.

**Comment by JustDalek** [comment]
I very much would be also interested in the number of invites users can generate per month.

As it stands right now, any user that is given this permission can fully undermine the intent of limiting the permission to certain people.

There's an argument about only giving the permissions to people you trust, but what if you dont trust people?

It's the same reason I give people their own credentials with their own permissions to admin/moderator  things instead of a general admin account I just give to people I trust. 

We love enforced limitations! 👏👏

**Comment by BakaDalek** [comment]
I wonder if there's any interest from devs on this feature. Currently the invite permissions completely undermines application only servers, it would be so nice to allow people to bring like, 3 friends with them and that's it ya know? 

Also global limits on how long invite links can live

**Comment by Tiwy57** [comment]
Hello,

Due to recent changes in the code, I was unable to easily override the "No Limit" directive in the form, which could have helped curb the phenomenon described. That's why I find this idea appealing.

It would allow for a more comprehensive configuration of the invitation system (which is admittedly somewhat archaic) and help limit the creation of infinite invitations, which could lead to moderation and potential security issues on invitation-only instances.

Additionally, administrators would no longer have to regularly monitor the creation of infinite invitations. This approach also encourages users to be more responsible, as they would need to specify a defined duration and a limited number of invitations, for example.

Thank you for your time.

**Comment by johannab33** [comment]
Hello - community Admin rather than coder here, but I've been following the invitation-control related issues that I was aware of for some time. I note the earlier one, which Renchap referenced above, was closed without being merged, which I presume means "not being done". Unsure if that indicates the idea of any better invitation control is done with, or it was just closing a duplicate. Will there be any continuation of this idea? I would really like to have my users able to invite, with the auditing of the "who invited who" trail, but not when the default is to create unlimited uses/no expiry.
